torchvision/datasets/_stereo_matching.py

```diff
@@ -0,0 +1,1041 @@
+        super().__init__(root=root)
+        self.transforms = transforms
+
+        self._images: List[Tuple[str, str]] = []
+        self._disparities: List[Tuple[str, str]] = []
+
+    def _read_img(self, file_path: str) -> Image.Image:
+        img = Image.open(file_path)
+        if img.mode != "RGB":
+            img = img.convert("RGB")
+        return img
+
+    def _scan_pairs(
+        self, paths_left_pattern: str, paths_right_pattern: str, fill_empty: bool = False
+    ) -> List[Tuple[str, str]]:
+        left_paths: List[str] = sorted(glob(paths_left_pattern))
+        right_paths: List[str] = sorted(glob(paths_right_pattern))
+
+        # used when dealing with inexistent disparity for the right image
+        if fill_empty:
+            right_paths = list("" for _ in left_paths)
+
+        if not left_paths:
+            raise FileNotFoundError(f"Could not find any files matching the patterns: {paths_left_pattern}")
+
+        if not right_paths:
+            raise FileNotFoundError(f"Could not find any files matching the patterns: {paths_right_pattern}")
+
+        if len(left_paths) != len(right_paths):
+            raise ValueError(
+                f"Found {len(left_paths)} left files but {len(right_paths)} right files using:\n "
+                f"left pattern: {paths_left_pattern}\n"
+                f"right pattern: {paths_right_pattern}\n"
+            )
+
+        images = list((left, right) for left, right in zip(left_paths, right_paths))
+        return images
+
+    @abstractmethod
+    def _read_disparity(self, file_path: str) -> Tuple:
+        # function that returns a disparity map and an occlusion map
+        pass
+
+    def __getitem__(self, index: int) -> Tuple:
+        """Return example at given index.
+
+        Args:
+            index(int): The index of the example to retrieve
+
+        Returns:
+            tuple: A 4-tuple with ``(img_left, img_right, disparity, valid_mask)`` where ``valid_mask``
+                is a numpy boolean mask of shape (H, W)
+                indicating which disparity values are valid. The disparity is a numpy array of
+                shape (1, H, W) and the images are PIL images. ``disparity`` and ``valid_mask`` are None for
+                datasets on which for ``split="test"`` the authors did not provide annotations.
+        """
+        img_left = self._read_img(self._images[index][0])
+        img_right = self._read_img(self._images[index][1])
+
+        dsp_map_left, valid_mask_left = self._read_disparity(self._disparities[index][0])
+        dsp_map_right, valid_mask_right = self._read_disparity(self._disparities[index][1])
+
+        imgs = (img_left, img_right)
+        dsp_maps = (dsp_map_left, dsp_map_right)
+        valid_masks = (valid_mask_left, valid_mask_right)
+
+        if self.transforms is not None:
+            (
+                imgs,
+                dsp_maps,
+                valid_masks,
+            ) = self.transforms(imgs, dsp_maps, valid_masks)
```

Comment:
Flattening everything would result in having to write very verbose function headers for transforms `(left_img, right_img, left_disp, right_disp, left_mask, right_mask)`.

I don't have any strong feelings towards this. The trade-off for tuples is that you have to do Tuple unpacking / manipulation inside the transform.

Reply:
I think for the current transform it will be kinda temporary (put in reference like optical_flow), and I have no strong opinion on this.

For the future transform API, I think it will be good to package left and right together (either stacked as tensor or maybe tuple). And the reason is because the new transforms accept `*input` as shown: https://github.com/pytorch/vision/blob/main/torchvision/prototype/transforms/_transform.py#L22  and in this case we can't apply transform separately on left and right image.